content/non-blocking-communication-pt1.rst

```diff
@@ -1,33 +1,307 @@
+synchronization is problematic when the time to send messages is
+comparable with either the amounts of compute work involved, or the
+variation in those amounts.
+
+Performant non-blocking MPI workflows for stencil applications
+--------------------------------------------------------------
+
+.. figure:: img/non-blocking-stylestencilworkflow.svg
+   :align: center
+
+   A correct non-blocking-style parallel stencil-computation
+   workflow. The MPI implementation may be able to do the
+   communication overlapped with the computation.
+
+This will also always work. It may be more efficient than the
+ring-style approach, but that depends on the implementation and
+environment, as well as the size of messages and amounts of compute
+that can participate in the overlapped region.
+
+The user is responsible for not using the memory area until after the
+message request has been waited upon.
+
+Non-blocking MPI send calls
+---------------------------
+
+An |term-MPI_Isend| is creates a send request and returns a request
+object. It may or may not have sent the message, or buffered it. The
+caller is responsible for not changing the buffer until after waiting
+upon the resulting request object.
+
+Call signature:
+
+.. code-block:: c
+
+    int MPI_Isend(const void* buf, int count, MPI_Datatype datatype,
+                  int dest, int tag, MPI_Comm comm, MPI_Request *request)
+
+.. note::
+
+    ``buf``, ``count`` and ``datatype`` describe the buffer to be sent
+    to ``dest`` rank of ``comm`` using tag ``tag``. The ``request`` object
+    that is returned must be used to wait on the communication later.
+
+Other calls exist for other sending modes, including buffered,
+synchronous, and ready-mode sends.
+
+
+Non-blocking MPI receive call
+-----------------------------
 
+An |term-MPI_Irecv| creates a receive request and returns a request
+object.  The caller is responsible for not changing the buffer until
+after waiting upon the resulting request object.
 
-Topic introduction here
+Call signature:
 
-You really want to browse this page alongside the source of it, to see
-how this is implemented.  See the links at the to right of the page.
+.. code-block:: c
 
+    int MPI_Irecv(void* buf, int count, MPI_Datatype datatype,
+                  int source, int tag, MPI_Comm comm, MPI_Request *request)
 
 
+.. note::
+
+    ``buf``, ``count`` and ``datatype`` describe the buffer to be
+    received from ``source`` rank of ``comm`` using tag ``tag``. The
+    ``request`` object that is returned must be used to wait on the
+    communication later.
+
+An |term-MPI_Irecv| can be used to match any kind of send, regardless of
+sending mode or blocking status.
+
+Waiting for non-blocking call completion
+----------------------------------------
+
+An |term-MPI_Wait| call waits for completion of the operation that
+created the request object passed to it. For a send, the semantics of
+the sending mode have been restored (not necessarily that the message
```

Comment:
I don't understand what "the semantics of the sending mode have been restored" :disappointed:

Reply:
You're right, I meant more like "fulfilled"